### ISSUE TITLE
Cryptography and Hash functions

### ISSUE THREAD

**Issue by leongross** [Rust-for-Linux/linux#684]
As far as I see it there is support for some cryptographic operations like [sha256 asm wrappers](https://rust-for-linux.github.io/docs/core/arch/x86_64/fn._mm_sha256msg1_epu32.html) or [single aes rounds](https://rust-for-linux.github.io/docs/core/arch/x86/fn._mm_aesdec_si128.html), but there are no higher level Hash Digest or Encryption functionalities comparable to [the crypto api](https://kernel.readthedocs.io/en/sphinx-samples/crypto-API.html#code-examples).

Are there plans to implement such interfaces? I searched the Github Issues for keywords but did not find any regarding hashing or encryption.

**Comment by bjorn3** [comment]
> As far as I see it there is support for some cryptographic operations like [sha256 asm wrappers](https://rust-for-linux.github.io/docs/core/arch/x86_64/fn._mm_sha256msg1_epu32.html) or [single aes rounds](https://rust-for-linux.github.io/docs/core/arch/x86/fn._mm_aesdec_si128.html),

These are part of the rust standard library and directly use the respective cpu instructions on x86/x86_64. They don't work on any other archs and don't work on older cpu's without those instructions. They probably shouldn't be directly used by your code.

> Are there plans to implement such interfaces?

Wrappers around linux kernel interfaces are mostly implemented once needed for drivers. Is there a specific driver you want to implement that needs certain crypto operations or are you just curious. If the former you could try implementing it yourself or see if someone else is willing to implement it. If the later we will eventually get support for it. It may just take a while. In any case keeping this issue open as a reminder to implement it is fine I think.

**Comment by ojeda** [comment]
Indeed, the crypto API of the kernel should be used (with a suitable wrapper).

**Comment by leongross** [comment]
> Is there a specific driver you want to implement that needs certain crypto operations or are you just curious

Yes I am implementing a kernel module that needs hashing functionality. 

> If the former you could try implementing it yourself or see if someone else is willing to implement it. If the later we will eventually get support for it

Is there any documentation on how to approach such kind of development? I have only seen the [rust-out-of-tree-module](https://github.com/Rust-for-Linux/rust-out-of-tree-module) project and the [Documentation/rust](https://github.com/Rust-for-Linux/linux/tree/rust/Documentation/rust).
But from these it is not clear to me how to integrate C headers in the build process (which seems to be essential to me) and which structural necessities have to be fulfilled.
A procedural guide for first-time contributors would be much appreciated and could assist engagement into the project!

**Comment by bjorn3** [comment]
I believe you have to add the respective C header to https://github.com/Rust-for-Linux/linux/blob/rust/rust/kernel/bindings_helper.h and then add a wrapper to the kernel rust crate. Rust-bindgen then automatically handles generating rust function declarations and type definitions from the C headers.

**Comment by ojeda** [comment]
> Is there any documentation on how to approach such kind of development? I have only seen the [rust-out-of-tree-module](https://github.com/Rust-for-Linux/rust-out-of-tree-module) project and the [Documentation/rust](https://github.com/Rust-for-Linux/linux/tree/rust/Documentation/rust).
But from these it is not clear to me how to integrate C headers in the build process (which seems to be essential to me) and which structural necessities have to be fulfilled.
A procedural guide for first-time contributors would be much appreciated and could assist engagement into the project!

You need to create the necessary abstractions in the `kernel` crate, and then use those from your module. In other words, it is not that you call or integrate C headers from your Rust module, which I think is what you are thinking about (given you mention the out-of-tree one).

At the moment, the best approach is to take a look at how other abstractions are written and how other modules use those abstractions. Eventually, we can write a guide on that. However, what we can already do is expand a bit on the "Abstractions vs. bindings" section and put some links to the relevant files (e.g. like the one @bjorn3 just gave you). I will take a look at that for next round.

**Comment by cqs21** [comment]
My driver also needs some crypto API (AEAD and skcipher), and I'm working on it right now. Last week, I committed a version of scatterlist wrapper: [cqs21/rust-for-linux@29fab3eb](https://github.com/cqs21/rust-for-linux/commit/29fab3eb37285f67ba6bc8b5d4e12122fb4bf82f). How could I contribute it to the `rust-next` branch? PR or mailing lists?

**Comment by ojeda** [comment]
> How could I contribute it to the `rust-next` branch? PR or mailing lists?

Please see https://rust-for-linux.com/contributing and https://rust-for-linux.com/branches.
